torch/_dynamo/compiled_autograd.py

```diff
@@ -0,0 +1,171 @@
+        self.compiler_fn = compiler_fn
+        self.stack = contextlib.ExitStack()
+        self.shape_env = ShapeEnv()
+        self.fake_tensor_mode = FakeTensorMode(
+            allow_fallback_kernels=True,
+            allow_non_fake_inputs=True,
+            shape_env=self.shape_env,
+        )
+        self.fx_tracer = PythonKeyTracer()
+        self.proxy_mode = ProxyTorchDispatchMode(self.fx_tracer, "symbolic")
+        self.hooks_proxy = None
+
+    def wrap_fake(self, x, source):
+        assert isinstance(x, torch.Tensor)
+        return self.fake_tensor_mode.from_tensor(x, source=source)
+
+    @staticmethod
+    def source(name, idx):
+        return GetItemSource(LocalSource(name), idx)
+
+    def begin_capture(self, inputs: List[torch.Tensor], sizes: List[int]):
+        counters["compiled_autograd"]["captures"] += 1
+        self.fx_tracer.root = torch.nn.Module()
+        self.fx_tracer.graph = torch.fx.Graph(tracer_cls=PythonKeyTracer)
+        self.fx_tracer.tensor_attrs = {}
+        args_proxy = self.fx_tracer.create_proxy("placeholder", "inputs", (), {})
+        sizes_proxy = self.fx_tracer.create_proxy("placeholder", "sizes", (), {})
+        self.hooks_proxy = self.fx_tracer.create_proxy("placeholder", "hooks", (), {})
+
+        # tensor inputs to fake tensors
+        inputs = [
+            self.wrap_fake(x, self.source("inputs", idx))
+            for idx, x in enumerate(inputs)
+        ]
+        proxies = [args_proxy[i] for i in range(len(inputs))]
+        self.bind_tensors_to_proxies(inputs, proxies)
+
+        # size inputs to symints
+        sizes = [
+            self.shape_env.create_symint_and_symbol(
+                val,
+                self.source("sizes", idx),
+                DimDynamic.DYNAMIC,
+            )
+            for idx, val in enumerate(sizes)
+        ]
+        self.bind_tensors_to_proxies(sizes, sizes_proxy)
+
+        # TODO(jansel): are all these modes needed?
+        self.stack.enter_context(decompose({}))
+        self.stack.enter_context(self.fake_tensor_mode)
+        self.stack.enter_context(self.proxy_mode.sym_mode)
+        self.stack.enter_context(self.proxy_mode)
+        self.stack.enter_context(disable_autocast_cache())
+        self.stack.enter_context(disable_proxy_modes_tracing(enable_current=True))
+        return inputs, sizes
+
+    def proxy_call_hook(self, hook, *args):
+        return self.fx_tracer.create_proxy(
+            "call_function",
+            call_hook,
+            (
+                hook,
+                *[self.to_proxy(x) for x in args],
+            ),
+            {},
+        )
+
+    def tensor_pre_hook(self, inputs, hook_id, i: int):
+        hook = self.hooks_proxy[hook_id]
+        proxy = self.proxy_call_hook(
+            hook,
+            inputs[i],
+        )
+        with disable_proxy_modes_tracing():
+            inputs[i] = inputs[i].clone()
+            self.bind_tensors_to_proxies([inputs[i]], [proxy])
+        return inputs
```

Comment:
I'm still a little shaky on how this particular part works. When exactly is this callback called?

Reply:
This callback is called during tracing at the moment in the graph when a tensor_pre_hook would be invoked. Unlike pre_hooks, tensor_pre_hook just takes a single input (rather than all inputs), but are otherwise the same.

This doesn't actually run the hook. The hook will be passed as an arg to the FX graph at call time. It just inserts a call_function node.